File: packages/auths-node/lib/org.ts
import native from './native'
import { mapNativeError, OrgError } from './errors'
import type { Auths } from './client'

/** Result of creating an organization. */
export interface OrgResult {
  /** Internal prefix for the organization. */
  orgPrefix: string
  /** The organization's KERI DID. */
  orgDid: string
  /** Human-readable label. */
  label: string
  /** Path to the registry storing the organization. */
  repoPath: string
}

/** An organization member record. */
export interface OrgMember {
  /** Delegated `did:keri:` AID minted by the org for this member. */
  memberDid: string
  /** Role within the organization (e.g. `'admin'`, `'member'`). */
  role: string
  /** Capabilities granted to this member. */
  capabilities: string[]
  /** DID of the organization that owns this membership. */
  orgDid: string
  /** KERI prefix of the member's delegated identity. */
  memberPrefix: string
  /** Whether the membership has been revoked. */
  revoked: boolean
  /** Expiration timestamp (RFC 3339), or `null` if no expiry. */
  expiresAt: string | null
}

/**
 * Check whether an organization member has admin role.
 *
 * @param member - The organization member to check.
 * @returns `true` if the member's role is `'admin'`.
 */
export function isAdmin(member: OrgMember): boolean {
  return member.role === 'admin'
}

/** Options for {@link OrgService.create}. */
export interface CreateOrgOptions {
  /** Human-readable label for the organization. */
  label: string
  /** Override the client's repo path. */
  repoPath?: string
  /** Override the client's passphrase. */
  passphrase?: string
}

/** Options for {@link OrgService.addMember}. */
export interface AddOrgMemberOptions {
  /** DID of the organization. */
  orgDid: string
  /** Human-readable alias for the member key minted by the org. */
  memberLabel: string
  /** Role to assign (e.g. `'admin'`, `'member'`). */
  role: string
  /** Capabilities to grant the member. If omitted, role defaults are used. */
  capabilities?: string[]
  /** Override the client's passphrase. */
  passphrase?: string
  /** Optional Unix timestamp (seconds) at which the membership expires. */
  expiresAt?: number
}

/** Options for {@link OrgService.revokeMember}. */
export interface RevokeOrgMemberOptions {
  /** DID of the organization. */
  orgDid: string
  /** Delegated DID of the member to revoke. */
  memberDid: string
  /** Override the client's passphrase. */
  passphrase?: string
}

/** Options for {@link OrgService.listMembers}. */
export interface ListOrgMembersOptions {
  /** DID of the organization. */
  orgDid: string
  /** Whether to include revoked members. Defaults to `false`. */
  includeRevoked?: boolean
}

/**
 * Manages organizations and their membership.
 *
 * The organization mints a fresh delegated key for each member from a label;
 * callers do not provide member public keys.
 *
 * Access via {@link Auths.orgs}.
 *
 * @example
 * ```typescript
 * const org = auths.orgs.create({ label: 'my-team' })
 * auths.orgs.addMember({
 *   orgDid: org.orgDid,
 *   memberLabel: 'alice',
 *   role: 'member',
 * })
 * ```
 */
export class OrgService {
  constructor(private client: Auths) {}

  /**
   * Creates a new organization.
   *
   * @param opts - Organization options.
   * @returns The created organization.
   * @throws {@link OrgError} if creation fails.
   *
   * @example
   * ```typescript
   * const org = auths.orgs.create({ label: 'engineering' })
   * console.log(org.orgDid) // did:keri:...
   * ```
   */
  create(opts: CreateOrgOptions): OrgResult {
    const rp = opts.repoPath ?? this.client.repoPath
    const pp = opts.passphrase ?? this.client.passphrase
    try {
      return native.createOrg(opts.label, rp, pp)
    } catch (err) {
      throw mapNativeError(err, OrgError)
    }
  }

  /**
   * Adds a member to an organization. The org mints a fresh delegated key for
   * the member from `memberLabel`; the returned `memberDid` is the new AID.
   *
   * @param opts - Member options.
   * @returns The new member record.
   * @throws {@link OrgError} if the operation fails.
   *
   * @example
   * ```typescript
   * const member = auths.orgs.addMember({
   *   orgDid: org.orgDid,
   *   memberLabel: 'alice',
   *   role: 'member',
   * })
   * ```
   */
  addMember(opts: AddOrgMemberOptions): OrgMember {
    const pp = opts.passphrase ?? this.client.passphrase
    const capsJson = opts.capabilities ? JSON.stringify(opts.capabilities) : null
    try {
      const result = native.addOrgMember(
        opts.orgDid,
        opts.memberLabel,
        opts.role,
        this.client.repoPath,
        capsJson,
        pp,
        opts.expiresAt ?? null,
      )
      return this.toMember(result)
    } catch (err) {
      throw mapNativeError(err, OrgError)
    }
  }

  /**
   * Revokes a member's access to an organization.
   *
   * @param opts - Revocation options.
   * @returns The updated member record with `revoked: true`.
   * @throws {@link OrgError} if the operation fails.
   */
  revokeMember(opts: RevokeOrgMemberOptions): OrgMember {
    const pp = opts.passphrase ?? this.client.passphrase
    try {
      const result = native.revokeOrgMember(opts.orgDid, opts.memberDid, this.client.repoPath, pp)
      return this.toMember(result)
    } catch (err) {
      throw mapNativeError(err, OrgError)
    }
  }

  /**
   * Lists members of an organization.
   *
   * @param opts - List options.
   * @returns Array of member records.
   * @throws {@link OrgError} if the operation fails.
   *
   * @example
   * ```typescript
   * const members = auths.orgs.listMembers({ orgDid: org.orgDid })
   * console.log(members.length)
   * ```
   */
  listMembers(opts: ListOrgMembersOptions): OrgMember[] {
    try {
      const json = native.listOrgMembers(
        opts.orgDid,
        opts.includeRevoked ?? false,
        this.client.repoPath,
        this.client.passphrase,
      )
      const raw = JSON.parse(json) as Array<Record<string, unknown>>
      return raw.map((m) => ({
        memberDid: m.member_did as string,
        role: (m.role as string) ?? 'member',
        capabilities: (m.capabilities as string[]) ?? [],
        orgDid: opts.orgDid,
        memberPrefix: (m.member_prefix as string) ?? '',
        revoked: m.revoked as boolean,
        expiresAt: (m.expires_at as string | null) ?? null,
      }))
    } catch (err) {
      throw mapNativeError(err, OrgError)
    }
  }

  private toMember(result: import('./native').NapiOrgMember): OrgMember {
    return {
      memberDid: result.memberDid,
      role: result.role,
      capabilities: JSON.parse(result.capabilitiesJson || '[]'),
      orgDid: result.issuerDid,
      memberPrefix: result.attestationRid,
      revoked: result.revoked,
      expiresAt: result.expiresAt ?? null,
    }
  }
}

File: packages/auths-python/python/auths/org.py
"""Organization management — Stripe-style API."""
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Optional

from auths._native import (
    add_org_member as _add_org_member,
    create_org as _create_org,
    list_org_members as _list_org_members,
    revoke_org_member as _revoke_org_member,
)
from auths._client import _map_error
from auths._errors import OrgError


@dataclass
class Org:
    """An organization identity."""

    prefix: str
    """KERI prefix of the organization identity."""
    did: str
    """The organization's DID (`did:keri:...`)."""
    label: str
    """Human-readable organization name."""
    repo_path: str
    """Path to the identity repository."""

    def __repr__(self):
        return f"Org(did={self.did!r}, label={self.label!r})"


@dataclass
class OrgMember:
    """A member within an organization."""

    member_did: str
    """Delegated `did:keri:` AID minted by the org for this member."""
    role: str
    """Member role: `"admin"`, `"member"`, or `"readonly"`."""
    capabilities: list[str]
    """Capabilities granted to this member."""
    org_did: str
    """DID of the organization that owns this membership."""
    member_prefix: str
    """KERI prefix of the member's delegated identity."""
    revoked: bool
    """Whether this membership has been revoked."""
    expires_at: Optional[str]
    """ISO 8601 expiry timestamp, or None for non-expiring memberships."""

    @property
    def is_admin(self) -> bool:
        """Whether this member has admin role."""
        return self.role == "admin"

    def __repr__(self):
        status = " revoked" if self.revoked else ""
        return f"OrgMember(did={self.member_did!r}, role={self.role!r}{status})"


class OrgService:
    """Resource service for organization operations."""

    def __init__(self, client):
        self._client = client

    def create(
        self,
        label: str,
        repo_path: str | None = None,
        passphrase: str | None = None,
    ) -> Org:
        """Create a new organization identity.

        Args:
            label: Human-readable name for the org.
            repo_path: Override identity store path.
            passphrase: Override passphrase.

        Returns:
            Org with the KERI prefix, DID, and label.

        Raises:
            OrgError: If organization creation fails.
            KeychainError: If the keychain is locked or inaccessible.

        Examples:
            ```python
            org = client.orgs.create("my-team")
            ```
        """
        rp = repo_path or self._client.repo_path
        pp = passphrase or self._client._passphrase
        try:
            prefix, did, lbl, rpath = _create_org(label, rp, pp)
            return Org(prefix=prefix, did=did, label=lbl, repo_path=rpath)
        except (ValueError, RuntimeError) as exc:
            raise _map_error(exc, default_cls=OrgError) from exc

    def add_member(
        self,
        org_did: str,
        member_label: str,
        role: str = "member",
        capabilities: list[str] | None = None,
        repo_path: str | None = None,
        passphrase: str | None = None,
        expires_at: int | None = None,
    ) -> OrgMember:
        """Add a member to an organization.

        The organization mints a fresh delegated key for the member from the
        given label; the returned `member_did` is the new delegated AID.

        Args:
            org_did: The organization's DID (`did:keri:...`).
            member_label: Human-readable alias for the minted member key.
            role: One of `"admin"`, `"member"`, `"readonly"`.
            capabilities: Explicit capability list. If None, uses role defaults.
            expires_at: Optional Unix timestamp at which the membership expires.

        Returns:
            OrgMember with the minted member DID and membership details.

        Raises:
            OrgError: If the member cannot be added.

        Examples:
            ```python
            member = client.orgs.add_member(org.did, "alice", role="member")
            ```
        """
        rp = repo_path or self._client.repo_path
        pp = passphrase or self._client._passphrase
        caps_json = json.dumps(capabilities) if capabilities else None
        try:
            m_did, r, caps_str, o_did, prefix, revoked, expires = _add_org_member(
                org_did, member_label, role, rp, caps_json, pp, expires_at,
            )
            return OrgMember(
                member_did=m_did,
                role=r,
                capabilities=json.loads(caps_str) if caps_str else [],
                org_did=o_did,
                member_prefix=prefix,
                revoked=revoked,
                expires_at=expires,
            )
        except (ValueError, RuntimeError) as exc:
            raise _map_error(exc, default_cls=OrgError) from exc

    def revoke_member(
        self,
        org_did: str,
        member_did: str,
        repo_path: str | None = None,
        passphrase: str | None = None,
    ) -> OrgMember:
        """Revoke a member's authorization.

        Args:
            org_did: The organization's DID.
            member_did: The member's delegated DID to revoke.

        Returns:
            OrgMember with revoked status.

        Raises:
            OrgError: If the member cannot be revoked.

        Examples:
            ```python
            revoked = client.orgs.revoke_member(org.did, member.member_did)
            ```
        """
        rp = repo_path or self._client.repo_path
        pp = passphrase or self._client._passphrase
        try:
            m_did, r, caps_str, o_did, prefix, revoked, expires = _revoke_org_member(
                org_did, member_did, rp, pp,
            )
            return OrgMember(
                member_did=m_did,
                role=r,
                capabilities=json.loads(caps_str) if caps_str else [],
                org_did=o_did,
                member_prefix=prefix,
                revoked=revoked,
                expires_at=expires,
            )
        except (ValueError, RuntimeError) as exc:
            raise _map_error(exc, default_cls=OrgError) from exc

    def update_member(
        self,
        org_did: str,
        member_did: str,
        member_label: str,
        role: str | None = None,
        capabilities: list[str] | None = None,
        repo_path: str | None = None,
        passphrase: str | None = None,
    ) -> OrgMember:
        """Update a member by revoking the old delegation and minting a new one.

        Because the organization mints member keys, an update revokes the
        existing delegated DID and adds a fresh member under `member_label`.

        Args:
            org_did: The organization's DID.
            member_did: The existing delegated DID to revoke.
            member_label: Alias for the newly minted member key.
            role: New role. If None, defaults to `"member"`.
            capabilities: New capabilities. If None, uses role defaults.

        Returns:
            OrgMember for the freshly minted member.

        Raises:
            OrgError: If the member cannot be updated.

        Examples:
            ```python
            updated = client.orgs.update_member(
                org.did, member.member_did, "alice", role="admin"
            )
            ```
        """
        self.revoke_member(
            org_did, member_did,
            repo_path=repo_path, passphrase=passphrase,
        )
        return self.add_member(
            org_did, member_label, role=role or "member",
            capabilities=capabilities,
            repo_path=repo_path, passphrase=passphrase,
        )

    def list_members(
        self,
        org_did: str,
        include_revoked: bool = False,
        repo_path: str | None = None,
    ) -> list[OrgMember]:
        """List all members of an organization.

        Args:
            org_did: The organization's DID.
            include_revoked: If True, includes revoked members.

        Returns:
            List of OrgMember objects.

        Raises:
            OrgError: If the organization doesn't exist.

        Examples:
            ```python
            members = client.orgs.list_members(org.did)
            ```
        """
        rp = repo_path or self._client.repo_path
        try:
            members_json = _list_org_members(org_did, include_revoked, rp)
            raw = json.loads(members_json)
            return [
                OrgMember(
                    member_did=m["member_did"],
                    role=m.get("role") or "member",
                    capabilities=m.get("capabilities") or [],
                    org_did=org_did,
                    member_prefix=m.get("member_prefix", ""),
                    revoked=m["revoked"],
                    expires_at=m.get("expires_at"),
                )
                for m in raw
            ]
        except (ValueError, RuntimeError) as exc:
            raise _map_error(exc, default_cls=OrgError) from exc

    def get_member(
        self,
        org_did: str,
        member_did: str,
        repo_path: str | None = None,
    ) -> OrgMember | None:
        """Look up a specific member.

        Args:
            org_did: The organization's DID.
            member_did: The member's DID to look up.

        Returns:
            OrgMember if found, or None.

        Examples:
            ```python
            member = client.orgs.get_member(org.did, dev.did)
            ```
        """
        members = self.list_members(org_did, include_revoked=False, repo_path=repo_path)
        return next((m for m in members if m.member_did == member_did), None)
